homeassistant/components/purpleair/config_flow.py

```diff
@@ -253,3 +336,105 @@ async def async_step_user(
+            self.hass,
+            self.config_entry.data[CONF_API_KEY],
+            user_input[CONF_LATITUDE],
+            user_input[CONF_LONGITUDE],
+            user_input[CONF_DISTANCE],
+        )
+
+        if validation.errors:
+            return self.async_show_form(
+                step_id="add_sensor",
+                data_schema=async_get_coordinates_schema(self.hass),
+                errors=validation.errors,
+            )
+
+        self._flow_data[CONF_NEARBY_SENSOR_OPTIONS] = async_get_nearby_sensors_options(
+            validation.data
+        )
+
+        return await self.async_step_choose_sensor()
+
+    async def async_step_choose_sensor(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the selection of a sensor."""
+        if user_input is None:
+            options = self._flow_data.pop(CONF_NEARBY_SENSOR_OPTIONS)
+            return self.async_show_form(
+                step_id="choose_sensor",
+                data_schema=async_get_nearby_sensors_schema(options),
+            )
+
+        sensor_index = int(user_input[CONF_SENSOR_INDEX])
+
+        if sensor_index in self.config_entry.options[CONF_SENSOR_INDICES]:
+            return self.async_abort(reason="already_configured")
+
+        options = deepcopy({**self.config_entry.options})
+        options[CONF_LAST_UPDATE_SENSOR_ADD] = True
+        options[CONF_SENSOR_INDICES].append(sensor_index)
+        return self.async_create_entry(title="", data=options)
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Manage the options."""
+        return self.async_show_menu(
+            step_id="init",
+            menu_options=["add_sensor", "remove_sensor"],
+        )
+
+    async def async_step_remove_sensor(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Add a sensor."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="remove_sensor",
+                data_schema=async_get_remove_sensor_schema(
+                    async_get_remove_sensor_options(self.hass, self.config_entry)
+                ),
+            )
+
+        device_registry = dr.async_get(self.hass)
+        device_entry = cast(
+            dr.DeviceEntry, device_registry.async_get(user_input[CONF_SENSOR_DEVICE_ID])
+        )
+
+        removed_sensor_index = async_get_sensor_index(
+            self.hass, self.config_entry, device_entry
+        )
+
+        # Remove the sensor's device from the device registry:
+        device_registry.async_update_device(
+            device_entry.id, remove_config_entry_id=self.config_entry.entry_id
+        )
+
+        # Untrack the sensor index in config entry options:
+        new_entry_options = async_untrack_sensor_index(
+            self.config_entry, removed_sensor_index
+        )
```

Comment:
Would this make sense as a general purpose helper?

Reply:
What do you think of 53518b618fe4? That normalizes the logic to work for both the options flow and for deletion of a device.